### PR TITLE
Adding attributes to models

### PR DESCRIPTION
**Pull request by vimukthi-git**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Issue Link: #713

### CODE REVIEW

File: documents/model.go [hunk 1]
```diff
@@ -109,6 +109,12 @@ type Model interface {
 
 	// CollaboratorCanUpdate returns an error if indicated identity does not have the capacity to update the document.
 	CollaboratorCanUpdate(updated Model, collaborator identity.DID) error
+
+	// AddAttribute adds a custom attribute to the model with the given value. If an attribute with the given name already exists its updated.
```
Comment (charleenfei) [inline documents/model.go]: typo: already exists, it is updated

File: documents/coredocument_test.go [hunk 2]
```diff
@@ -488,3 +488,21 @@ func TestCoreDocument_GetSignCollaborators(t *testing.T) {
...
 }
+
+func TestCoreDocument_AddAttribute(t *testing.T) {
+	id1 := testingidentity.GenerateRandomDID()
+	cas := CollaboratorsAccess{ReadWriteCollaborators: []identity.DID{id1}}
+	cd, err := NewCoreDocumentWithCollaborators(nil, cas)
+	assert.NoError(t, err)
+	err = cd.AddAttribute("com.basf.deliverynote.chemicalnumber", Str, "100")
+	assert.NoError(t, err)
+	assert.True(t, len(cd.Attributes) == 1)
```
Comment (charleenfei) [inline documents/coredocument_test.go]: assert.Len?

File: documents/coredocument.go [hunk 3]
```diff
@@ -587,6 +590,30 @@ func (cd *CoreDocument) Timestamp() (time.Time, error) {
 	return utils.FromTimestamp(cd.Document.Timestamp)
 }
 
+// AddAttribute adds a custom attribute to the model with the given value. If an attribute with the given name already exists its updated.
```
Comment (charleenfei) [inline documents/coredocument.go]: typo: already exists, it's/ it is updated

File: documents/attribute.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+type AllowedAttributeType string
+
+// String string repr
+func (a AllowedAttributeType) String() string {
+	return string(a)
+}
+
+const (
+	// Int256 is the standard integer custom attribute type
+	Int256 AllowedAttributeType = "int256"
```
Comment (mikiquantum) [inline documents/attribute.go]: Can we use suffix/prefix `Type`? For example: `Int256Type` or `TypeInt256`

File: documents/coredocument_test.go (hunk 2 above)
Comment (charleenfei) [reply documents/coredocument_test.go]: could we also test cases with empty readableKey or nil value?

File: crypto/generate.go [hunk 5]
```diff
@@ -39,11 +39,17 @@ func GenerateSigningKeyPair(publicFileName, privateFileName, curveType string) (
...
 	h := sha256.New()
-	_, err = h.Write(preimage)
+	_, err = h.Write(value)
 	if err != nil {
-		return []byte{}, []byte{}, err
+		return []byte{}, err
 	}
-	hash = h.Sum(hash)
-	return preimage, hash, nil
+	hash = h.Sum(nil)
```
Comment (charleenfei) [inline crypto/generate.go]: why is this line necessary if we are just appending nil?

File: documents/coredocument_test.go (hunk 2 above)
Comment (vimukthi-git) [reply documents/coredocument_test.go]: As the `TODO` says below these would be updated when converters are ready. Right now its just a sanity check.

File: crypto/generate.go (hunk 5 above)
Comment (vimukthi-git) [reply crypto/generate.go]: yes we can just make the following line
`return h.Sum(nil), nil`

File: documents/coredocument.go [hunk 6]
```diff
@@ -587,6 +590,30 @@ func (cd *CoreDocument) Timestamp() (time.Time, error) {
...
+
+// GetAttribute gets the attribute with the given name from the model together with its type, it returns a non-nil error if the attribute doesn't exist or can't be retrieved.
+func (cd *CoreDocument) GetAttribute(name string) (hashedKey []byte, attrType string, value interface{}, valueStr string, err error) {
+	if attr, ok := cd.Attributes[name]; ok {
+		// TODO convert value to its string repr
+		return attr.hashedKey, string(attr.attrType), attr.value, "", nil
+	}
+	return hashedKey, attrType, value, valueStr, errors.NewTypedError(ErrCDAttribute, errors.New("attribute does not exist"))
+}
+
```
Comment (mikiquantum) [inline documents/coredocument.go]: Should we add a deleteAttribute?

File: crypto/generate.go (hunk 5 above)
Comment (charleenfei) [reply crypto/generate.go]: I mean, we can also just return the hash, no? Sum() does not change the underlying hash state, just appends the current hash to the parameter and returns.

File: documents/coredocument_test.go (hunk 2 above)
Comment (charleenfei) [reply documents/coredocument_test.go]: 👍

File: crypto/generate.go (hunk 5 above)
Comment (vimukthi-git) [reply crypto/generate.go]: As far as I know there is no other function to get the underlying hash anyway, so we have to call this.

**Comment by vimukthi-git** [comment]
@mikiquantum @charleenfei thanks for the reviews, I will address your concerns ASAP.
After this is merged we will park the feature temporarily for Amphitrite release. Since @lucasvo wants me to focus on other more critical stuff.

File: crypto/generate.go (hunk 5 above)
Comment (charleenfei) [reply crypto/generate.go]: okay, I see

File: documents/coredocument.go [hunk 7]
```diff
@@ -587,6 +590,36 @@ func (cd *CoreDocument) Timestamp() (time.Time, error) {
...
+// DeleteAttribute deletes a custom attribute from the model
+func (cd *CoreDocument) DeleteAttribute(name string) error {
```
Comment (vedhavyas) [inline documents/coredocument.go]: Any mutations to CoreDocument should return the new model. This will ensure that we dont mutate the current one
